### ISSUE TITLE
Provide a Reactive alternative to KafkaOperations and KafkaTemplate

### ISSUE THREAD

**Issue by sdeleuze** [spring-projects/spring-kafka#43]
[KafkaOperations](https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaOperations.java) and [KafkaTemplate](https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java) does not currently provide a way to perform truly non-blocking operations since they are using [`java.util.concurrent.Future`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html) where the only way to retrieve the result is a [blocking `get()`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#get%28%29) method.

Providing a Reactive alternative to these classes using [Reactive Streams](http://www.reactive-streams.org/) and [Reactor Core](https://spring.io/blog/2016/03/11/reactor-core-2-5-becomes-a-unified-reactive-foundation-on-java-8) types, like in our new [Reactive HTTP client](https://github.com/spring-projects/spring-reactive/issues/13#issuecomment-163975472) (which is a Reactive alternative to `RestTemplate`), in the Reactive Spring Data work that is about to start (see this [ReactiveMongoOperations](https://github.com/mp911de/spring-data-mongodb/blob/7c758318b09106b4d36bb0d4927606f6ad9f771b/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/ReactiveMongoOperations.java#L71) draft) or in the [new Cloud Foundry Java client](https://github.com/cloudfoundry/cf-java-client#usage) would enable truly async non-blocking and Reactive usage of `spring-kafka` and would be a huge improvement in term of scalability, especially when used with Reactive Spring Web or Reactive Spring Data support, or even with the [upcoming Reactive support in Spring Cloud Stream 1.1](https://github.com/spring-cloud/spring-cloud-stream/issues/458).

Based on all the work we have done on this topic on [Spring Reactive](https://github.com/spring-projects/spring-reactive) to prepare the Reactive support of Spring Framework 5, I would advise to use:
- raw Reactive Streams `Publisher` for input parameter to accept any Reactive Streams implementation like Reactor `Mono` / `Flux`, Akka Streams types or upcoming RxJava 2 [`Flowable`](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/Flowable.java) type.
- [Reactor `Mono`](https://projectreactor.io/core/docs/api/reactor/core/publisher/Mono.html) return type for async single values (it implements Reactive Streams `Publisher`, can be easily converted to a blocking value with `get()`, to a Java 8 `CompletableFuture` or to RxJava `Single`).
- [Reactor `Flux`](https://projectreactor.io/core/docs/api/reactor/core/publisher/Flux.html) for streams that will emit multiple values and where we want to be able to subscribe (it also implements Reactive Streams `Publisher`, can be easily concerted into RxJava types).

I am available with @smaldini and the Spring Reactive team to help bridging Kafka client API with `Flux` and `Mono` types. I think it should be doable (and efficient)  by using Kafka 0.9 [new Producer API](http://kafka.apache.org/documentation.html#producerapi) and [new Consumer API](http://kafka.apache.org/documentation.html#newconsumerapi).

**Comment by artembilan** [comment]
@sdeleuze ,

Thank you for the interest for this project and your energy to share the Reactive experience everywhere!

So, I don't want to say that you should provide the solution here or even answer immediately (just with your convenience), but anyway I'd like to discuss this one. Just because it isn't clear what you are going to do here.

1. `KafkaTemplate` is fully based on the  Kafka 0.9 Producer API:
````java
Future<RecordMetadata> future;
if (this.producerListener == null) {
	future = this.producer.send(producerRecord);
}
````
where their `Future` is just like `FutureRecordMetadata` implementation.

2. `KafkaTemplate` provides only `send` operations. Yes, that's why it is only `Kafka Producer` based.

3. Since `Future` can be simple adapted by the [`Mono`](https://projectreactor.io/core/docs/api/reactor/core/publisher/Mono.html#fromFuture-java.util.concurrent.Future-), I don't so much benefits with just moving the single code line `Mono.fromFuture(this.producer.send(producerRecord));` into the `KafkaTemplate` instead of doing that from the target user's code.

Just curious what is in your mind on the matter. For better understanding the entire `Flux` and `Mono` story.

Thank you one more time!

**Comment by sdeleuze** [comment]
Hey @artembilan,

That's my pleasure to help if I can, and indeed I am very interested in using Kafka in Spring applications!

I think Kafka 0.9 producer API and documentation can be **very** confusing and I think I understand now better why `KafkaTemplate` returns a `Future`.

[KafkaProducer](http://kafka.apache.org/090/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#send(org.apache.kafka.clients.producer.ProducerRecord)) API has 2 `send()` variants:
1.  `Future<RecordMetadata> send(ProducerRecord<K,V> record)`
2.  `Future<RecordMetadata> send(ProducerRecord<K,V> record, Callback callback)`

The first variant documentation says "Asynchronously send a record to a topic." This is not false, but **very** misleading since as soon as you call `future.get()` to be able to get the `RecordMetadata` that will block the current thread. Even if you convert it later to another truly async type like `Flux`, `Mono` or `CompletableFuture`, getting the result of a `Future` will always be blocking.

It took me some times to understand how we could use `KafkaProducer` in a truly async non-blocking way. The answer seems to be in the 2nd variant Javadoc : you should do nothing with the `Future` return value and instead retreive the metadata thanks to the [callback parameter](http://kafka.apache.org/090/javadoc/org/apache/kafka/clients/producer/Callback.html). This kind of API is extremely confusing, IMO having the second variant returning no value at all (`void`) would have been far less confusing.

So back to the initial topic, I think `Mono` and `Flux` should be created using the callback parameter of the second `send()` variant, without using the `Future` return value at all. I guess something like [`Mono.fromCallable()`](https://projectreactor.io/core/docs/api/reactor/core/publisher/Mono.html#fromCallable-java.util.concurrent.Callable-) could be used, but to be validated by @smaldini.

For a better understanding of the full `Mono` and `Flux` story, maybe [this comment](https://github.com/spring-cloud/spring-cloud-deployer/pull/19#issuecomment-203624278) I wrote could give you more context about the various stages we went through while working on Spring Reactive.

That's was not the original purpose of this issue, but for the current async `KafkaTemplate` API my proposal would be to expose a callback or [ListenableFuture](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/util/concurrent/ListenableFuture.html) based API (using internaly the 2nd variant + the callback parameter) rather than the current `Future` based approach that is blocking (no choice) as soon as you get the metadatas.

Any thoughts?

**Comment by garyrussell** [comment]
Hi @sdeleuze - notice that the template has a `ProducerListener` abstraction that allows the user to ignore the `Future` and get a callback instead - via the `ProducerListenerInvokingCallback `, we provide more contextual data than Kafka alone (sent payload etc).

By default, a [simple logging callback is used](https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java#L52), which just logs errors.

[Code here](https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java#L204).

**Comment by sdeleuze** [comment]
Thanks for pointing me to that @garyrussell.

So indeed that allows to register a common global listener, but still forces users to perform a blocking operation if they want to do something specific with the metadata on each `send()` invocation right ?

**Comment by artembilan** [comment]
@sdeleuze ,

thank you for your investigation and clues!
Really appreciate :smile: 

I wonder you mean something like this: https://github.com/artembilan/spring-kafka/commit/11385d638a797c14a2d57f393fd52f4fc24d5622

Pay attention that I use `MonoProcessor` because I wasn't able to find anything else for deferred `Mono` completion.
From other side pay attention, please, to the additional `testMono()`.
I tried to `consume()` the `Mono` from the `reactiveSend()` the same way, but looks like, if we complete `MonoProcessor` from different Thread (like in case of the `KafkaProducer`), the consumption is skipped when `onNext()` happens before `consume()`. 

Any ideas?

CC @smaldini

**Comment by smaldini** [comment]
@artembilan when you're around we can get on screenhero :+1: Sounds like a bug, MonoProcessor is basically what used to be called Promise and is effectively the only way to send at most event arbitrary (like the FluxProcessor components you played with).

**Comment by garyrussell** [comment]
Just FYI; we are planning the 1.0 release for late April early May.

I don't believe we can require Java 8 for that version.

Perhaps we should develop the 2.0 version in parallel?

**Comment by sdeleuze** [comment]
Indeed I thought that may be a blocking point for you, so why not:
 - A 1.0 based on [`ListenableFuture`](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/util/concurrent/ListenableFuture.html) which is truly non blocking instead of `Future` (it would use internally the callback approach to provide a real async non-blocking behavior)
 - A 2.0 based on Spring 5 master (so Java 8+) that is using `Mono` instead of `ListenableFuture`, and maybe `Flux` where that makes sense

**Comment by artembilan** [comment]
Hello, guys!

Thank you for your activity and sorry for delay. This is just a late morning for me :smile: 

So, regarding Java 8. We can just can introduce a new `ReactiveKafkaTemplate` class and mark it with `@UsesJava8`.

But I agree about the `2.0` story and replacing `ListenableFuture`. We just need to figure out what's wrong with my solution :smile: 

BTW, I wasn't able to reproduce it with this test-case:
````java
	@Test
	public void testMono() throws Exception {
		MonoProcessor<String> promise = MonoProcessor.create();
		final CountDownLatch onNextCountDownLatch = new CountDownLatch(1);
		Executors.newSingleThreadExecutor().execute(() -> {
			promise.onNext("test");
			onNextCountDownLatch.countDown();
		});
		assertThat(onNextCountDownLatch.await(10, TimeUnit.SECONDS)).isTrue();

		final CountDownLatch successCountDownLatch = new CountDownLatch(1);
		promise.consume(v -> successCountDownLatch.countDown());
		assertThat(successCountDownLatch.await(10, TimeUnit.SECONDS)).isTrue();
	}
````
so, my assumption regarding multi-threading may be wrong...

Adding the `Thread.sleep(1000);` to the `Callback` before `recordMetadataMonoProcessor.onNext(metadata);` makes my test happy.

What is bad here with Reactor (again) it is difficult to debug :cry: .

So, @smaldini , I'd be glad to Screenhero with you on the matter! 
Feel free to ping me when you have spare time

**Comment by artembilan** [comment]
After some Screenhero session with @smaldini we localized the problem in the `MonoProcesson`, but not clear yet what is the race condition, when we `subscribe` just a bit late after `onNext`:
```
1459439786969 ON_NEXT: Thread[kafka-producer-network-thread | producer-1,5,main]
ConsumerRecord(topic = templateTopic, partition = 0, offset = 3, key = 22, value = Mono)
1459439786977 ADDED CONSUMER
```

**Comment by garyrussell** [comment]
@sdeleuze - here's my first cut at an `AsyncKafkaTemplate` (1.0) returning `ListenableFuture`.

Let me know if you have any comments/suggestions before I complete the tests and docs.

**Comment by garyrussell** [comment]
@sdeleuze - here's my first cut at an `AsyncKafkaTemplate` (1.0) returning `ListenableFuture`.

Let me know if you have any comments/suggestions before I complete the tests and docs.

https://github.com/spring-projects/spring-kafka/pull/54

**Comment by sdeleuze** [comment]
@garyrussell I think we should avoid to expose an API that returns raw `Future` because [this is a trap](http://stackoverflow.com/questions/34213272/kafka-producer-0-9-performance-issue-with-small-messages/34257254#34257254). I like the original intent to provide an async API (that make perfectly sense for Kafka) in `KafkaTemplate`, but with `Future` you end up with a sync behavior as soon as you want to know when the future is complete or get the value it emits.

Since `ListenableFuture`  implements `Future`, my proposal would be to change methods returning `Future` in `KafkaTemplate` to methods returning `ListenableFuture` using the implementation you have in your PR. And the good thing with `ListenableFuture` is that it still implements `Future.get()` that allows to provide a sync behavior, that's why I think you could also remove all the `Kafka.sync*()` methods, and just explain in the Javadoc that users can call `Kafka.send("foo", "bar").get()` if they want such behavior (IMO this should be only used for test or demo).

That would also enable us to add easily a `ReactiveKafkaTemplate` without requiring a 2.0 (we just have to add an optional dependency to Reactor Core using @artembilan work) where `Mono` is used instead of `ListenableFuture`, and maybe with some `Flux` based methods to handle that in a streaming way.

That would make a pretty nice and clear `spring-kafka` public API where `KafkaTemplate` provides a real async non-blocking API, that you can easily turn into a sync one by using `get()`, and a Reactive API that could be introduced when we want (1.1 maybe ?) as an additional `ReactiveKafkaTemplate` that won't break `KafkaTemplate` API provided in 1.0.

Any thoughts?

**Comment by garyrussell** [comment]
@sdeleuze Sorry - yes, I should have made it clear that my intention was to replace the existing template with this one, which I just hacked together as a PoC to see if it was in line with your thoughts.

**Comment by sdeleuze** [comment]
Oh in that case perfect :-) And yes the implementation is exactly what I had in mine with the use of `SettableListenableFuture` + Kafka `Producer` callback 👍

**Comment by garyrussell** [comment]
Updated PR here: https://github.com/spring-projects/spring-kafka/pull/57

**Comment by sdeleuze** [comment]
As discussed previously, Kafka Streams is going to be shipped with Kafka 0.10.0, as detailed in [this great blog post](http://www.confluent.io/blog/introducing-kafka-streams-stream-processing-made-simple).

While I think both `KStream` (Kafka Streams) and `Flux` (Reactor) will have there own advantages and are interesting to support, one open question we will have to solve is : should we build Reactive Streams support directly based on low level Kafka `Producer` / `Consumer` API or on top of Kafka Streams by  converting `KStream` to Reactor `Flux` for example ...

**Comment by mbogoevici** [comment]
As far as I can tell, the `KStream` interface is mainly designed with a functional programming model in mind, but not necessarily reactive pull. That it follows a push model driven by the consumer thread becomes clearer when looking at https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L311. So, at a first glance, I do not believe that this fits well with building a Reactive Streams wrapper around it. I *do* see an interesting opportunity of hooking in `Flux`/`Mono` by adapting stream component pipelines, though so let's keep the discussion open.

**Comment by artembilan** [comment]
Just for the full picture: https://github.com/reactor/reactor-kafka

**Comment by adamdec** [comment]
Hi, do you plan to integrate reactor-kafka into spring-kafka?

**Comment by garyrussell** [comment]
Yes, that is the plan, after it stabilizes.

**Comment by PedroAlvarado** [comment]
@garyrussell Help me understand this better. Fast forward a number of releases, there will be built-in support for both kstreams and reactor-kafka. Moreover, the kstreams integration will not be wrapped around reactive-streams story. Correct?

**Comment by garyrussell** [comment]
Correct.

**Comment by lpandzic** [comment]
Is this targeted for 2.0?

**Comment by artembilan** [comment]
I don't think so.
The Spring Kafka 2.0 is targeted for the September 21, just after Spring Framework 5.0 GA
And I don't see any plans for Reactor Kafka GA release. 

@ilayaperumalg , any thoughts, please?

More over there is still need some work to do from our side on the matter.
So, in my best feeling this task is going to be deferred to the `2.1` version.

**Comment by rotilho** [comment]
Hello @artembilan! Are there still any plans for a reactive kafka operations?

**Comment by artembilan** [comment]
They are, indeed, but we haven't started `2.1` branch yet.

Any contribution is welcome, of course!

**Comment by hartmut-co-uk** [comment]
@artembilan meaning `2.2`? 
https://github.com/spring-projects/spring-kafka/releases

Is there any timeline for this?
Thanks.

**Comment by garyrussell** [comment]
We currently have higher priorities; community contributions would be most welcome.

**Comment by hartmut-co-uk** [comment]
Hi, many thanks for the update! 
..didn't mean to add pressure - I'll try to find time to have a closer look.
